File: src/algorithms/matching.ts
import { GraphData, AlgorithmExecutionResult, AlgorithmStep } from '../types/graph';
import { DistressCall, RescueTeam } from '../types/simulation';

export interface RescueMatch {
  teamId: string;
  teamName: string;
  distressId: string;
  targetNodeId: string;
  priority: string;
  distanceKm: number;
  estimatedEtaMin: number;
}

export interface ShelterAllocationMatch {
  residentialNodeId: string;
  residentialName: string;
  population: number;
  shelterNodeId: string;
  shelterName: string;
  travelDistanceKm: number;
}

// Euclidean distance helper
function getDistance(n1: { lat: number; lng: number }, n2: { lat: number; lng: number }): number {
  const dLat = (n2.lat - n1.lat) * 111.32;
  const dLng = (n2.lng - n1.lng) * 111.32 * Math.cos((n1.lat * Math.PI) / 180);
  return Math.sqrt(dLat * dLat + dLng * dLng);
}

// Bipartite Matching for Rescue Teams <-> Distress Requests
export function runRescueTeamMatching(
  graph: GraphData,
  teams: RescueTeam[],
  distressCalls: DistressCall[]
): { matches: RescueMatch[]; execution: AlgorithmExecutionResult } {
  const startTime = performance.now();
  const steps: AlgorithmStep[] = [];
  const matches: RescueMatch[] = [];

  const pendingCalls = distressCalls.filter(c => c.status === 'pending' || !c.assignedTeamId);
  const availableTeams = teams.filter(t => t.status === 'idle' || !t.assignedDistressId);

  steps.push({
    stepIndex: 0,
    description: `Initializing Bipartite Maximum Matching for ${availableTeams.length} available rescue units and ${pendingCalls.length} active emergency distress calls.`,
    highlightedNodeIds: pendingCalls.map(c => c.nodeId),
    highlightedEdgeIds: []
  });

  // Priority-weighted Greedy Matching / Hungarian reduction
  // Sort distress calls by priority: P1 (Critical) -> P2 (Urgent) -> P3 (Stable)
  const priorityScore = { P1: 3, P2: 2, P3: 1 };
  const sortedCalls = [...pendingCalls].sort((a, b) => priorityScore[b.priority] - priorityScore[a.priority]);

  const assignedTeams = new Set<string>();

  for (const call of sortedCalls) {
    const callNode = graph.nodes[call.nodeId];
    if (!callNode) continue;

    let bestTeam: RescueTeam | null = null;
    let minCost = Infinity;
    let bestDist = 0;

    for (const team of availableTeams) {
      if (assignedTeams.has(team.id)) continue;
      const teamNode = graph.nodes[team.currentNodeId];
      if (!teamNode) continue;

      const dist = getDistance(teamNode, callNode);
      // Cost incorporates distance and priority urgency
      const cost = dist / (priorityScore[call.priority] * 1.5);

      if (cost < minCost) {
        minCost = cost;
        bestTeam = team;
        bestDist = dist;
      }
    }

    if (bestTeam) {
      assignedTeams.add(bestTeam.id);
      const etaMin = Math.max(2, Math.round((bestDist / bestTeam.speedKmH) * 60));
      matches.push({
        teamId: bestTeam.id,
        teamName: bestTeam.name,
        distressId: call.id,
        targetNodeId: call.nodeId,
        priority: call.priority,
        distanceKm: Math.round(bestDist * 10) / 10,
        estimatedEtaMin: etaMin
      });

      if (steps.length < 30) {
        steps.push({
          stepIndex: steps.length,
          description: `Matched [${call.priority}] "${call.description}" at ${callNode.name} -> Team "${bestTeam.name}" (ETA: ~${etaMin} min, Dist: ${bestDist.toFixed(1)}km).`,
          highlightedNodeIds: [call.nodeId, bestTeam.currentNodeId],
          highlightedEdgeIds: []
        });
      }
    }
  }

  const elapsed = performance.now() - startTime;
  return {
    matches,
    execution: {
      algorithmName: 'Priority-Weighted Bipartite Matching for Rescue Dispatch',
      executionTimeMs: Math.round(elapsed * 100) / 100,
      summary: `Successfully matched ${matches.length} emergency rescue units to high-priority disaster distress calls.`,
      steps,
      customMetrics: {
        'Active Distress Calls': distressCalls.length,
        'Assigned Rescue Teams': matches.length,
        'Critical (P1) Coverage': `${matches.filter(m => m.priority === 'P1').length}/${distressCalls.filter(d => d.priority === 'P1').length}`,
        'Avg Response ETA': `${Math.round(matches.reduce((acc, m) => acc + m.estimatedEtaMin, 0) / Math.max(matches.length, 1))} min`
      }
    }
  };
}

// Capacity-Constrained Matching: Evacuee Neighborhoods to Optimal Shelters
export function runShelterAllocationMatching(graph: GraphData): {
  allocations: ShelterAllocationMatch[];
  execution: AlgorithmExecutionResult;
} {
  const startTime = performance.now();
  const steps: AlgorithmStep[] = [];
  const allocations: ShelterAllocationMatch[] = [];

  const residentialNodes = Object.values(graph.nodes).filter(n => n.type === 'residential' || n.hazardRisk > 0.3);
  const shelterNodes = Object.values(graph.nodes).filter(n => n.type === 'shelter' || n.type === 'hospital');

  const shelterRemainingCap: Record<string, number> = {};
  for (const s of shelterNodes) {
    shelterRemainingCap[s.id] = s.capacity || 2500;
  }

  steps.push({
    stepIndex: 0,
    description: `Initiated Capacity-Constrained Gale-Shapley Matching for ${residentialNodes.length} zones across ${shelterNodes.length} designated emergency shelters.`,
    highlightedNodeIds: shelterNodes.map(s => s.id),
    highlightedEdgeIds: []
  });

  for (const rNode of residentialNodes) {
    // Sort candidate shelters by distance
    const sortedShelters = [...shelterNodes].map(s => ({
      shelter: s,
      distance: getDistance(rNode, s)
    })).sort((a, b) => a.distance - b.distance);

    let assigned = false;
    for (const { shelter, distance } of sortedShelters) {
      if (shelterRemainingCap[shelter.id] >= rNode.population * 0.4) {
        shelterRemainingCap[shelter.id] -= rNode.population;
        allocations.push({
          residentialNodeId: rNode.id,
          residentialName: rNode.name,
          population: rNode.population,
          shelterNodeId: shelter.id,
          shelterName: shelter.name,
          travelDistanceKm: Math.round(distance * 10) / 10
        });
        assigned = true;

        if (steps.length < 35) {
          steps.push({
            stepIndex: steps.length,
            description: `Allocated ${rNode.population} residents from "${rNode.name}" -> Shelter "${shelter.name}" (${distance.toFixed(1)}km). Remaining Cap: ${Math.max(0, shelterRemainingCap[shelter.id])}.`,
            highlightedNodeIds: [rNode.id, shelter.id],
            highlightedEdgeIds: []
          });
        }
        break;
      }
    }
  }

  const elapsed = performance.now() - startTime;
  return {
    allocations,
    execution: {
      algorithmName: 'Capacity-Constrained Optimal Shelter Matching',
      executionTimeMs: Math.round(elapsed * 100) / 100,
      summary: `Allocated ${allocations.length} residential hazard zones to nearest available shelters without overloading capacity limits.`,
      steps,
      customMetrics: {
        'Zones Allocated': allocations.length,
        'Total Evacuees Accommodated': allocations.reduce((acc, a) => acc + a.population, 0).toLocaleString(),
        'Total Shelters Utilized': shelterNodes.filter(s => (shelterRemainingCap[s.id] < (s.capacity || 2500))).length,
        'Avg Evacuee Distance': `${(allocations.reduce((acc, a) => acc + a.travelDistanceKm, 0) / Math.max(allocations.length, 1)).toFixed(1)} km`
      }
    }
  };
}

File: backend/algorithms/matching.py
import math
import time
from typing import Dict, List, Any, Tuple
from ..models.graph import GraphData
from ..models.simulation import DistressCall, RescueTeam
from ..models.algorithm_results import AlgorithmExecutionResult, AlgorithmStep

def get_distance(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    d_lat = (lat2 - lat1) * 111.32
    d_lng = (lng2 - lng1) * 111.32 * math.cos(math.radians(lat1))
    return math.sqrt(d_lat * d_lat + d_lng * d_lng)

def run_rescue_team_matching(
    graph: GraphData,
    teams: List[RescueTeam],
    distress_calls: List[DistressCall]
) -> Tuple[List[Dict[str, Any]], AlgorithmExecutionResult]:
    start_time = time.perf_counter()
    steps: List[AlgorithmStep] = []
    matches: List[Dict[str, Any]] = []

    pending_calls = [c for c in distress_calls if c.status == 'pending' or not c.assignedTeamId]
    available_teams = [t for t in teams if t.status == 'idle' or not t.assignedDistressId]

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Initialized Priority Bipartite Maximum Matching for {len(available_teams)} available rescue units and {len(pending_calls)} active SOS distress calls.",
        highlightedNodeIds=[c.nodeId for c in pending_calls],
        highlightedEdgeIds=[]
    ))

    priority_weight = {'P1': 3, 'P2': 2, 'P3': 1}
    sorted_calls = sorted(pending_calls, key=lambda c: priority_weight.get(c.priority, 1), reverse=True)
    assigned_teams = set()

    for call in sorted_calls:
        call_node = graph.nodes.get(call.nodeId)
        if not call_node:
            continue

        best_team = None
        min_cost = float('inf')
        best_dist = 0.0

        for team in available_teams:
            if team.id in assigned_teams:
                continue
            team_node = graph.nodes.get(team.currentNodeId)
            if not team_node:
                continue

            dist = get_distance(team_node.lat, team_node.lng, call_node.lat, call_node.lng)
            cost = dist / (priority_weight.get(call.priority, 1) * 1.5)
            if cost < min_cost:
                min_cost = cost
                best_team = team
                best_dist = dist

        if best_team:
            assigned_teams.add(best_team.id)
            eta_min = max(2, int((best_dist / max(best_team.speedKmH, 20.0)) * 60))
            match_data = {
                'teamId': best_team.id,
                'teamName': best_team.name,
                'distressId': call.id,
                'targetNodeId': call.nodeId,
                'priority': call.priority,
                'distanceKm': round(best_dist, 1),
                'estimatedEtaMin': eta_min
            }
            matches.append(match_data)

            if len(steps) < 30:
                steps.append(AlgorithmStep(
                    stepIndex=len(steps),
                    description=f"Matched [{call.priority}] SOS '{call.description}' at {call_node.name} -> Team '{best_team.name}' (ETA: ~{eta_min}m, Dist: {best_dist:.1f}km).",
                    highlightedNodeIds=[call.nodeId, best_team.currentNodeId],
                    highlightedEdgeIds=[]
                ))

    elapsed = (time.perf_counter() - start_time) * 1000

    avg_eta = sum(m['estimatedEtaMin'] for m in matches) / max(len(matches), 1)

    return matches, AlgorithmExecutionResult(
        algorithmName='Priority-Weighted Bipartite Matching for Rescue Dispatch (Python)',
        executionTimeMs=round(elapsed, 2),
        summary=f"Dispatched {len(matches)} specialized emergency rescue units to high-priority disaster distress calls.",
        steps=steps,
        customMetrics={
            'Active Distress Calls': len(distress_calls),
            'Assigned Rescue Teams': len(matches),
            'Critical (P1) Coverage': f"{sum(1 for m in matches if m['priority'] == 'P1')}/{sum(1 for d in distress_calls if d.priority == 'P1')}",
            'Avg Response ETA': f"{round(avg_eta)} min",
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )

def run_shelter_allocation_matching(
    graph: GraphData
) -> Tuple[List[Dict[str, Any]], AlgorithmExecutionResult]:
    start_time = time.perf_counter()
    steps: List[AlgorithmStep] = []
    allocations: List[Dict[str, Any]] = []

    residential_nodes = [n for n in graph.nodes.values() if n.type == 'residential' or n.hazardRisk > 0.3]
    shelter_nodes = [n for n in graph.nodes.values() if n.type in ('shelter', 'hospital')]

    shelter_rem_cap = {s.id: (s.capacity if s.capacity else 2500) for s in shelter_nodes}

    steps.append(AlgorithmStep(
        stepIndex=0,
        description=f"Initiated Capacity-Constrained Gale-Shapley Matching for {len(residential_nodes)} hazard zones across {len(shelter_nodes)} designated emergency sanctuaries.",
        highlightedNodeIds=[s.id for s in shelter_nodes],
        highlightedEdgeIds=[]
    ))

    for r_node in residential_nodes:
        # Sort shelters by distance
        sorted_shelters = sorted(
            shelter_nodes,
            key=lambda s: get_distance(r_node.lat, r_node.lng, s.lat, s.lng)
        )

        for shelter in sorted_shelters:
            dist = get_distance(r_node.lat, r_node.lng, shelter.lat, shelter.lng)
            if shelter_rem_cap[shelter.id] >= r_node.population * 0.4:
                shelter_rem_cap[shelter.id] -= r_node.population
                alloc = {
                    'residentialNodeId': r_node.id,
                    'residentialName': r_node.name,
                    'population': r_node.population,
                    'shelterNodeId': shelter.id,
                    'shelterName': shelter.name,
                    'travelDistanceKm': round(dist, 1)
                }
                allocations.append(alloc)

                if len(steps) < 35:
                    steps.append(AlgorithmStep(
                        stepIndex=len(steps),
                        description=f"Allocated {r_node.population} evacuees from '{r_node.name}' -> Sanctuary '{shelter.name}' ({dist:.1f}km). Rem Cap: {max(0, shelter_rem_cap[shelter.id])}.",
                        highlightedNodeIds=[r_node.id, shelter.id],
                        highlightedEdgeIds=[]
                    ))
                break

    elapsed = (time.perf_counter() - start_time) * 1000

    total_evac = sum(a['population'] for a in allocations)
    avg_dist = sum(a['travelDistanceKm'] for a in allocations) / max(len(allocations), 1)

    return allocations, AlgorithmExecutionResult(
        algorithmName='Capacity-Constrained Optimal Shelter Matching (Python)',
        executionTimeMs=round(elapsed, 2),
        summary=f"Allocated {len(allocations)} residential communities ({total_evac:,} evacuees) to closest safe shelters without capacity overflow.",
        steps=steps,
        customMetrics={
            'Zones Allocated': len(allocations),
            'Total Evacuees Accommodated': f"{total_evac:,}",
            'Shelters Utilized': sum(1 for s in shelter_nodes if shelter_rem_cap[s.id] < (s.capacity or 2500)),
            'Avg Evacuee Distance': f"{avg_dist:.1f} km",
            'Execution Speed': f"{round(elapsed, 2)} ms"
        }
    )
